### PR TITLE
Move broadcast of key updates into sync, fixup of externaljwt generation / test

### LINKED ISSUE #128672: Issue with protobindings codegen
Hello, ever since yesterday with https://github.com/kubernetes/kubernetes/pull/128190 merged I've been having some issues with codegen for protobindings. Not sure if there's a configuration that I might be missing, want to see if anyone else has encountered.

```
kubernetes ❯ ./hack/update-codegen.sh protobindings 
+++ [1107 18:14:20] Generating protobuf bindings for 9 targets
google/protobuf/timestamp.proto: File not found.
api.proto:24:1: Import "google/protobuf/timestamp.proto" was not found or had errors.
api.proto:77:5: "google.protobuf.Timestamp" is not defined.
!!! [1107 18:14:34] Call tree:
!!! [1107 18:14:34]  1: ~/gospace/src/k8s.io/kubernetes/hack/lib/protoc.sh:40 kube::protoc::protoc(...)
!!! [1107 18:14:34]  2: hack/_update-generated-proto-bindings-dockerized.sh:50 kube::protoc::generate_proto(...)
!!! [1107 18:14:34] Call tree:
!!! [1107 18:14:34]  1: ~/gospace/src/k8s.io/kubernetes/hack/lib/protoc.sh:40 kube::protoc::protoc(...)
!!! [1107 18:14:34]  2: hack/_update-generated-proto-bindings-dockerized.sh:50 kube::protoc::generate_proto(...)
!!! [1107 18:14:34] Call tree:
!!! [1107 18:14:34]  1: ./hack/update-codegen.sh:888 codegen::protobindings(...)
```

```
kubernetes took 15s ❯ which protoc
~/gospace/src/k8s.io/kubernetes/third_party/protoc/protoc
kubernetes ❯ protoc --version  
libprotoc 23.4
```

/cc @liggitt @HarshalNeelkamal

### PR DESCRIPTION
**Pull request by liggitt**
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

1. Bugfix: Moves broadcast into syncKeys method, so any change to keys (periodic or on-demand) always notifies listeners immediately.
2. Test flake fix: Address race between setting new keys and ensuring they've been read (https://github.com/kubernetes/kubernetes/pull/128190#issuecomment-2462857707)
3. Add proto definition file needed for local generation (fixes https://github.com/kubernetes/kubernetes/issues/128672)

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

/assign @enj
/sig auth

### CODE REVIEW

File: pkg/serviceaccount/externaljwt/plugin/keycache.go [hunk 1]
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
...
+		return true
+	}
+	// Avoid deepequal checks on key content itself.
+	// If the number of keys changed, we changed
+	if len(oldPublicKeys.Keys) != len(newPublicKeys.Keys) {
+		return true
+	}
+	// If the order, key id, or oidc discovery flag changed, we changed.
+	for i := range oldPublicKeys.Keys {
+		if oldPublicKeys.Keys[i].KeyID != newPublicKeys.Keys[i].KeyID {
```
Comment (HarshalNeelkamal) [inline pkg/serviceaccount/externaljwt/plugin/keycache.go:183]: shall we short before doing this?

Comment (liggitt) [reply pkg/serviceaccount/externaljwt/plugin/keycache.go:183]: no, if the backend changes the key order, I'm fine propagating that as a change

File: pkg/serviceaccount/externaljwt/plugin/keycache.go [hunk 2]
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
...
 
 		p.verificationKeys.Store(newPublicKeys)
-
 		externaljwtmetrics.RecordKeyDataTimeStamp(newPublicKeys.DataTimestamp.Unix())
 
+		if keysChanged(oldPublicKeys, newPublicKeys) {
+			p.broadcastUpdate()
+		}
+
 		return newPublicKeys, nil
```
Comment (enj) [inline pkg/serviceaccount/externaljwt/plugin/keycache.go:165]:
```suggestion
		return nil, nil
```

File: pkg/serviceaccount/externaljwt/plugin/keycache.go [hunk 3]
```diff
@@ -152,35 +146,57 @@ func (p *keyCache) findKeyForKeyID(keyID string) ([]serviceaccount.PublicKey, bo
...
+	// If the timestamp changed, we changed
+	if !oldPublicKeys.DataTimestamp.Equal(newPublicKeys.DataTimestamp) {
+		return true
+	}
+	// Avoid deepequal checks on key content itself.
+	// If the number of keys changed, we changed
+	if len(oldPublicKeys.Keys) != len(newPublicKeys.Keys) {
+		return true
+	}
+	// If the order, key id, or oidc discovery flag changed, we changed.
```
Comment (enj) [inline pkg/serviceaccount/externaljwt/plugin/keycache.go:181]: `bytes.Equal` compare the keys?

**Review by enj — COMMENTED** [review]
Minor comments.

File: pkg/serviceaccount/externaljwt/plugin/keycache.go (hunk 2 above)
Comment (liggitt) [reply pkg/serviceaccount/externaljwt/plugin/keycache.go]: fixed

File: pkg/serviceaccount/externaljwt/plugin/keycache.go (hunk 3 above)
Comment (liggitt) [reply pkg/serviceaccount/externaljwt/plugin/keycache.go:181]: we don't have bytes at this point, we have opaque instantiated public keys, I wasn't sure about this bit... I think the data timestamp and key count / order / logical ids is fine

**Comment by liggitt** [comment]
addressed comments, also folded in distinct follow-up commits to deflake an integration flake @benluddy observed and make proto generation work locally

**Pull request merged**